document-portal/document-portal-fuse.c

```diff
@@ -1897,7 +1899,25 @@ xdp_fuse_open (fuse_req_t req,
     return;
 
   path = fd_to_path (inode->physical->fd);
-  fd = open (path, open_flags, 0);
+  if (open_flags & O_NOFOLLOW)
+    {
+      int res;
+      char actual_path[PATH_MAX + 1];
+      /*
+       * `path` is a path to the fd entry in `/proc`, which is a symlink
+       * to the actual file. Opening it directly with `O_NOFOLLOW` will
+       * fail. So we should resolve it first then we can honour the no
+       * follow flag.
+       */
+      res = readlink (path, actual_path, sizeof (actual_path));
+      if (res == -1)
+        return xdp_reply_err (op, req, errno);
+      actual_path[res] = '\0';
```

Comment:
I think this should perform more validation checks for the output of `readlink`. As a reference, [this is what the BPF kernel tool does](https://github.com/torvalds/linux/blob/master/tools/bpf/bpftool/common.c#L387-L395).

Reply:
Actually the fix is to pass `sizeof (actual_path) - 1` as we make sure it's null terminated.

Reply:
... and still checking res for the length because if it truncated then it's wrong. and return `ENOENT` as errno.

Reply:
fixed a bit more thoroughly